compiler/rustc_target/src/abi/mod.rs

```diff
@@ -1031,19 +1031,22 @@ impl Niche {
         assert!(bits <= 128);
         let max_value = !0u128 >> (128 - bits);
 
-        if count > max_value {
-            return None;
-        }
-
-        // Compute the range of invalid values being reserved.
-        let start = v.end.wrapping_add(1) & max_value;
-        let end = v.end.wrapping_add(count) & max_value;
-
-        if v.contains(end) {
+        let niche = v.end.wrapping_add(1)..v.start;
+        let available = niche.end.wrapping_sub(niche.start) & max_value;
+        if count > available {
             return None;
         }
 
-        Some((start, Scalar { value, valid_range: v.with_end(end) }))
+        let (start, valid_range) = if v.start <= max_value - v.end && v.start >= count {
```

Comment:
I guess what I was missing was the idea that it "preferred closeness (i.e. the smaller space)" as opposed to "preferring the larger space".

Reply:
I am going to unmark this as resolved only for the fact that I think the *intention* needs to be in the comment, since I don't think it's obvious from the code without already having been told the intention.

This could also be partly done by using explanative variables, e.g.:
```rust
let dist_to_0_below = v.start;
let dist_to_0_above = max_value - v.end + 1;

// Prefer the direction that reaches `0` quicker.
if dist_to_0_below < dist_to_0_above && v.start >= count {...}
```

But, what if it's a wraparound range? The "distance to 0" stuff becomes meaningless, since `0` is a valid value, and can never be reached. I guess `v.start >= count` isn't required for correctness, because of the earlier check, but it still feels weird to not have `v.start < v.end` as a condition.

Reply:
I understand your concern about `v.start < v.end`.
It is possible to just add it in the if expression, I just feel bad about it when it is not having any influence on the outcome.
I added a more thorough explanation and introduced some helper variables. 
https://github.com/rust-lang/rust/pull/87794/commits/4e1f80cb6a75c8f81acc5292b6245ff03cd07f54

Reply:
I have commited https://github.com/rust-lang/rust/pull/87794/commits/52f4be95a8d488b60543993a526fbc3429b65ec8
Changed so it checks for `v.start > v.end`.
It avoids moving past zero for `v.end`